src/Microsoft.AspNetCore.Http/StreamPipeWriter.cs

```diff
@@ -0,0 +1,313 @@
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncInternal(CancellationToken cancellationToken = default)
+        {
+            if (_bytesWritten == 0)
+            {
+                return new FlushResult(isCanceled: false, IsCompletedOrThrow());
+            }
+
+            try
+            {
+                CancellationTokenRegistration reg;
+                if (cancellationToken.CanBeCanceled)
+                {
+                    reg = cancellationToken.Register(state => ((StreamPipeWriter)state).Cancel(), this);
+                    using (reg)
+                    {
+                        return await WriteAndFlushAsync();
+
+                    }
+                }
+                else
+                {
+                    return await WriteAndFlushAsync();
+                }
+            }
+            finally
+            {
+                _bytesWritten = 0;
+                _position = 0;
+            }
+        }
+
+        private async ValueTask<FlushResult> WriteAndFlushAsync()
+        {
+            try
+            {
+                // Write all completed segments and whatever remains in the current segment
+                // and flush the result.
+                if (_completedSegments != null && _completedSegments.Count > 0)
+                {
+                    var count = _completedSegments.Count;
+                    for (var i = 0; i < count; i++)
+                    {
+                        var segment = _completedSegments[i];
+#if NETCOREAPP2_2
+                        await _writingStream.WriteAsync(segment.Buffer.Slice(0, segment.Length), InternalTokenSource.Token);
+#elif NETSTANDARD2_0
+                        await _writingStream.WriteAsync(segment.Buffer.ToArray(), 0, segment.Length, InternalTokenSource.Token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                        segment.Return();
+                    }
+
+                    _completedSegments.Clear();
+                }
+
+                if (!_currentSegment.IsEmpty)
+                {
+#if NETCOREAPP2_2
+                    await _writingStream.WriteAsync(_currentSegment.Slice(0, _position), InternalTokenSource.Token);
+#elif NETSTANDARD2_0
+                    await _writingStream.WriteAsync(_currentSegment.ToArray(), 0, _position, InternalTokenSource.Token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                }
+
+                await _writingStream.FlushAsync(_internalTokenSource.Token);
+            }
+            catch (OperationCanceledException)
+            {
+                // Remove the cancellation token such that the next time Flush is called
+                // A new CTS is created.
+                Interlocked.Exchange(ref _internalTokenSource, null);
+
+                // Catch any cancellation and translate it into setting isCanceled = true
+                return new FlushResult(isCanceled: true, IsCompletedOrThrow());
```

Comment:
We need to throw if `cancellationToken` that was passed in FlushAsync() was cancelled.